README.md
# format-thousands

[![Commitizen friendly][commitizen-image]][commitizen-url]
[![XO code style][codestyle-image]][codestyle-url]

[![NPM version][npm-image]][npm-url]
[![Build Status][build-image]][build-url]
[![Coveralls Status][coveralls-image]][coveralls-url]
[![Dependency Status][depstat-image]][depstat-url]
[![DevDependency Status][depstat-dev-image]][depstat-dev-url]

> Format thousands with custom separator: 1 000 000

Demo: [vovanr.github.io/format-thousands][demo]

## Install

```shell
npm install --save format-thousands
```

## Usage

```js
var formatThousands = require('format-thousands');

formatThousands(1000);
//=> '1 000'

formatThousands(5000, {formatFourDigits: false});
//=> '5000'

formatThousands(1000000, '`');
//=> '1`000`000'

formatThousands(10000, {separator: "'"});
//=> "10'000"

formatThousands(-100000);
//=> "-100 000"

formatThousands(10000.0001)
//=> "10 000.0001"

formatThousands();
//=> ''
```

## License
MIT © [Vladimir Rodkin](https://github.com/VovanR)

[demo]: https://vovanr.github.io/format-thousands

[commitizen-url]: https://commitizen.github.io/cz-cli/
[commitizen-image]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?style=flat-square

[codestyle-url]: https://github.com/xojs/xo
[codestyle-image]: https://img.shields.io/badge/code_style-XO-5ed9c7.svg?style=flat-square

[npm-url]: https://npmjs.org/package/format-thousands
[npm-image]: https://img.shields.io/npm/v/format-thousands.svg?style=flat-square

[build-url]: https://github.com/VovanR/format-thousands/actions?query=workflow%3A%22Tests%22
[build-image]: https://img.shields.io/github/workflow/status/VovanR/format-thousands/Tests?style=flat-square

[coveralls-url]: https://coveralls.io/r/VovanR/format-thousands
[coveralls-image]: https://img.shields.io/coveralls/VovanR/format-thousands.svg?style=flat-square

[depstat-url]: https://david-dm.org/VovanR/format-thousands
[depstat-image]: https://david-dm.org/VovanR/format-thousands.svg?style=flat-square

[depstat-dev-url]: https://david-dm.org/VovanR/format-thousands
[depstat-dev-image]: https://david-dm.org/VovanR/format-thousands/dev-status.svg?style=flat-square


CHANGELOG.md
<a name="2.0.0"></a>
# [2.0.0](https://github.com/VovanR/format-thousands/compare/v1.1.0...v2.0.0) (2020-12-30)


### Bug Fixes

* **lib:** Fix decimal and negative formatting ([3cac3ef](https://github.com/VovanR/format-thousands/commit/3cac3ef)), closes [#7](https://github.com/VovanR/format-thousands/issues/7)


### build

* **package:** Upgrade development dependencies ([495a1fd](https://github.com/VovanR/format-thousands/commit/495a1fd))


### BREAKING CHANGES

* **package:** Drop support of Node.js < v10



<a name="1.1.1"></a>
## [1.1.1](https://github.com/VovanR/format-thousands/compare/v1.1.0...v1.1.1) (2017-04-09)


### Bug Fixes

* **lib:** Fix decimal and negative formatting ([4a3c779](https://github.com/VovanR/format-thousands/commit/4a3c779)), closes [#7](https://github.com/VovanR/format-thousands/issues/7)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/vovanr/format-thousands/compare/v1.0.4...v1.1.0) (2017-02-16)


### Features

* **lib:** Add `formatFourDigits` option ([1f8c15a](https://github.com/vovanr/format-thousands/commit/1f8c15a)), closes [#5](https://github.com/vovanr/format-thousands/issues/5)



<a name="1.0.4"></a>
## [1.0.4](https://github.com/vovanr/format-thousands/compare/v1.0.3...v1.0.4) (2016-04-28)


### Bug Fixes

* **lib:** Fix bug when number 0 formatted to ''([8cf8388](https://github.com/vovanr/format-thousands/commit/8cf8388))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/vovanr/format-thousands/compare/v1.0.2...v1.0.3) (2016-04-21)



test.js
const test = require('ava');
const fn = require('.');

const NBSP = String.fromCharCode(160);

test('is function', t => {
  t.is(typeof fn, 'function');
});

test('should return `1`', t => {
  t.is(fn(1), '1');
});

test('should return `0`', t => {
  t.is(fn(0), '0');
});

test('should return `12`', t => {
  t.is(fn(12), '12');
});

test('should return `123`', t => {
  t.is(fn(123), '123');
});

test('should return `1 000`', t => {
  t.is(fn(1000), `1${NBSP}000`);
});

test('should return `10 000`', t => {
  t.is(fn(10000), `10${NBSP}000`);
});

test('should set default thousands separator', t => {
  t.is(fn(1234567890), `1${NBSP}234${NBSP}567${NBSP}890`);
});

test('should set custom thousands separator', t => {
  t.is(fn(1234567890, '_'), '1_234_567_890');
  t.is(fn(1234567890, {separator: '_'}), '1_234_567_890');
});

test('should not set separator if length < 5', t => {
  const options = {formatFourDigits: false};
  t.is(fn(1000, options), '1000');
  t.is(fn(10000, options), `10${NBSP}000`);
});

test('should set separator if length < 5', t => {
  const options = {formatFourDigits: true};
  t.is(fn(1000, options), `1${NBSP}000`);
  t.is(fn(10000, options), `10${NBSP}000`);
});

test('should return empty string if fired without args', t => {
  t.is(fn(), '');
});

test('should not format `null` and `undefined`', t => {
  t.is(fn(null), '');
  t.is(fn(undefined), '');
});

test('should not format frac', t => {
  t.is(fn(10000.0001, ' '), '10 000.0001');
  t.is(fn(-1000000000.545, ' '), '-1 000 000 000.545');
  t.is(fn(123456.1234567, ' '), '123 456.1234567');
});

test('should not format sign', t => {
  t.is(fn(-100000, ' '), '-100 000');
  t.is(fn(-100000000.545, ' '), '-100 000 000.545');
});


index.js
/**
 * Non-Breaking Space `&nbsp;`
 * @const {string}
 */
const NBSP = String.fromCharCode(160);

/**
 * @typedef {Object} ParsedNumber
 * @prop {string} integer
 * @prop {string} fraction
 * @prop {string} sign
 */

/**
 * @param {number} number
 * @returns {ParsedNumber}
 */
function parseNumber(number) {
  const isNegative = number < 0;
  let numberString = String(number);

  if (isNegative) {
    numberString = numberString.slice(1);
  }

  const decimal = numberString.split('.');

  return {
    integer: decimal[0],
    fraction: decimal[1] || '',
    sign: isNegative ? '-' : ''
  };
}

/**
 * @param {number} number
 * @param {string} separator
 * @returns {string}
 */
function format(number, separator) {
  number = String(number);

  while (number.length % 3) {
    number = '#' + number;
  }

  let result = number.substr(0, 3);
  result = result.replace(/#/g, '');

  let i;
  const {length} = number;
  for (i = 3; i < length; i += 3) {
    result = result + separator + number.substr(i, 3);
  }

  return result;
}

/**
 * @param {number} number
 * @param {Object|string} [options=" "]
 * @param {string} [options.separator=" "]
 * @param {boolean} [options.formatFourDigits=true]
 * @returns {string}
 *
 * @example
 * formatThousands(1000);
 * //=> '1 000'
 *
 * formatThousands(5000, {formatFourDigits: false});
 * //=> '5000'
 *
 * formatThousands(10000, {separator: "'"});
 * //=> "10'000"
 */
module.exports = function (number, options) {
  let result = '';
  let separator = NBSP;
  let formatFourDigits = true;

  if (!number && number !== 0) {
    return result;
  }

  const numberObject = parseNumber(number);
  const numberString = String(number);

  if (typeof options === 'object') {
    if (options.separator) {
      ({separator} = options);
    }

    if (typeof options.formatFourDigits === 'boolean') {
      ({formatFourDigits} = options);
    }
  } else if (typeof options !== 'undefined') {
    separator = options;
  }

  if (
    numberObject.integer.length <= 3 ||
    (numberObject.integer.length === 4 && !formatFourDigits)
  ) {
    result = numberString;
  } else {
    result += numberObject.sign;
    result += format(numberObject.integer, separator);
    if (numberObject.fraction) {
      result += '.';
      result += numberObject.fraction;
    }
  }

  return result;
};
